src/array_api_typing/_array.py

```diff
@@ -67,10 +74,131 @@ def dtype(self, /) -> DTypeT_co:
         ...
 
 
+class HasDevice(Protocol):
+    """Protocol for array classes that have a device attribute."""
+
+    @property
+    def device(self) -> object:  # TODO: more specific type
+        """Hardware device the array data resides on."""
+        ...
+
+
+class HasMatrixTranspose(Protocol):
```

Comment:
I'd move this one closer to `HasTranspose`

Reply:
And the use-case of this is to require at least 2 dimensions, right?

Reply:
> This is the order defined in the array api docs.

Ah sure

> Don't most/all Array objects have this method and do that check?

I was talking about the static typing stuff, but numpy currently allows transposing 1d arrays, but the intention is to change that.